### ISSUE TITLE
sql: a query hangs on release-19.2 branch

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#43076]
This query hangs on `19.2` branch (I tried 19.2.1 and the current 19.2 branch):
```
SELECT
			sha1('\x847b':::BYTES::BYTES)::STRING AS col_6407,
			e'zp|&\f':::STRING AS col_6408,
			0.8016126389930045:::FLOAT8 AS col_6409,
			tab_3651.col_6402 AS col_6410,
			'MW)o!Is':::STRING:::NAME AS col_6411,
			tab_3653.col_6405 AS col_6412
		FROM
			(
				VALUES
					(0.013145047396657256:::FLOAT8, 'X':::STRING),
					(NULL, e'\x00':::STRING),
					('NaN':::FLOAT8, COALESCE('':::STRING:::NAME, 'xw':::STRING:::NAME)),
					((-1.9883442925077763):::FLOAT8, ';Qdk':::STRING:::NAME),
					(
						(1.3663809054956113:::FLOAT8::FLOAT8 // (-0.7725723048637168):::FLOAT8::FLOAT8)::FLOAT8,
						e'&bu\t\x07j\x1e':::STRING:::NAME
					),
					(1.5784056562638735:::FLOAT8, e'|$P\x1e-':::STRING:::NAME)
			)
				AS tab_3651 (col_6401, col_6402)
			INNER JOIN (
					VALUES
						(e'\x0e;\x1e':::STRING),
						((e'_\x1eKp`UD':::STRING::STRING || e'\x10':::STRING::STRING)::STRING),
						(
							(e'{"2H+9V}R,\'I0": [[]], "Hpl?,;v4=s": {"c": {"8>COjA}D": []}}, "XLwP": {"(;>a": false}, "b": {}, "bar": {"X|]7": {}, "bar": {"a": false, "foobar": []}, "vSLU": "f,4?n^\'$b )"}, "~cy": {}}':::JSONB::JSONB->>746654939041373451:::INT8::INT8)::STRING
						),
						(e'z\x03':::STRING),
						(
							('[{"c": {"8[by&5d>-Kc": "PIC6g75S`", "SRXHH{UJ^/S": {}}, "j*+[*O_>.`%": [], "xR#>": false}, {"b": {">D{Qj]op=": {}}, "baz": {"N{PN1DK@Szd": []}, "foo": true}]':::JSONB::JSONB->>NULL::INT8)::STRING
						),
						('ibw:OCE}':::STRING)
				)
					AS tab_3652 (col_6403) ON true
			JOIN (VALUES ('?h':::STRING, 28966:::INT8)) AS tab_3653 (col_6404, col_6405) ON
					(tab_3651.col_6402) = (tab_3653.col_6404)
			JOIN (
					VALUES
						(e'h\x04A%o=':::STRING),
						('':::STRING),
						(NULL),
						((NULL::JSONB->>e'\nGp':::STRING::STRING)::STRING),
						(NULL)
				)
					AS tab_3654 (col_6406) ON
					(tab_3651.col_6402) = (tab_3654.col_6406)
					AND (tab_3653.col_6404) = (tab_3654.col_6406)
					AND (tab_3652.col_6403) = (tab_3654.col_6406)
		ORDER BY
			tab_3653.col_6404;
```
Note that `EXPLAIN`ing the query also hangs and that on the current master (20.1) branch it succeeds.

**Comment by yuzefovich** [comment]
I debugged this a little, and the query is stuck here https://github.com/cockroachdb/cockroach/blob/release-19.2/pkg/sql/opt/xform/optimizer.go#L208, so it seems to be an issue with the optimizer, and I'm reassigning the issue to Radu for triage.

**Comment by RaduBerinde** [comment]
Could be the same with #42035. @rytaft do you think we can backport that fix?

**Comment by rytaft** [comment]
Opened #43082 to backport the fix. Confirmed that this query no longer hangs.

**Comment by rytaft** [comment]
Fixed by #43082
